compressors/arithmetic_coding.py

```diff
@@ -0,0 +1,436 @@
+        self.params = params
+
+        # define the probability model used by the AEC
+        # the model can get updated when we call update_model(s) after every step
+        self.freq_model = freq_model_cls(freq_base, params.MAX_ALLOWED_TOTAL_FREQ)
+
+    @classmethod
+    def shrink_range(cls, freqs: Frequencies, s: Any, low: int, high: int) -> Tuple[int, int]:
+        """shrinks the range (low, high) based on the symbol s
+
+        Args:
+            s (Any): symbol to encode
+
+        Returns:
+            Tuple[int, int]: (low, high) ranges returned after shrinking
+        """
+
+        # compute some intermediate variables: rng, c, d
+        rng = high - low
+        c = freqs.cumulative_freq_dict[s]
+        d = c + freqs.frequency(s)
+
+        # perform shrinking of low, high
+        # NOTE: this is the basic Arithmetic coding step implemented using integers
+        high = low + (rng * d) // freqs.total_freq
+        low = low + (rng * c) // freqs.total_freq
+        return (low, high)
+
+    def encode_block(self, data_block: DataBlock):
+        """Encode block function for arithmetic coding"""
+
+        # ensure data_block.size is not too big
+        err_msg = "choose a larget DATA_BLOCK_SIZE_BITS, as data_block.size is too big"
+        assert data_block.size < (1 << self.params.MAX_BLOCK_SIZE), err_msg
+
+        # initialize the low and high states
+        low = 0
+        high = self.params.FULL
+
+        # initialize the output
+        encoded_bitarray = BitArray("")
+
+        # add the data_block size at the beginning
+        # NOTE: Arithmetic decoding needs a way to indicate where to stop the decoding
+        # One way is to add a character at the end which signals EOF. This requires us to
+        # change the probabilities of the other symbols. Another way is to just signal the size of the
+        # block. These two approaches add a bit of overhead.. the approach we use is much more transparent
+        encoded_bitarray = uint_to_bitarray(data_block.size, self.params.DATA_BLOCK_SIZE_BITS)
+
+        # initialize counter for mid-range re-adjustments
+        num_mid_range_readjust = 0
+
+        # start the encoding
+        for s in data_block.data_list:
+
+            # ensure freqs.total_freq is not too big
+            err_msg = """the frequency total is too large, which might cause stability issues. 
+            Please increase the precision (or reduce the total_freq"""
+            assert (
+                self.freq_model.freqs_current.total_freq < self.params.MAX_ALLOWED_TOTAL_FREQ
+            ), err_msg
+
+            # shrink range
+            # i.e. the core Arithmetic encoding step
+            low, high = ArithmeticEncoder.shrink_range(self.freq_model.freqs_current, s, low, high)
+            ## update the freq model for encoding the next symbol
+            self.freq_model.update_model(s)
+
+            # perform re-normalizing range
```

Comment:
@kedartatwawadi: More comments can be added in this section unless you were thinking of explaining all this in detail in write-up. Given just the code, it's hard to decode the logic based on just the comments. I am not sure what should be our level of intention here with the comments.

Reply:
ll do